File: budgetsync-api/app/routers/transactions.py
from datetime import date
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from ..dependencies import CurrentUser, get_current_user, get_db
from ..schemas.transaction import TransactionBulkCreate, TransactionCreate, TransactionListResponse, TransactionRead, TransactionUpdate
from ..services.transactions import (
    create_transaction,
    create_transactions_bulk,
    delete_transaction,
    list_transactions,
    reset_transactions,
    update_transaction,
)

router = APIRouter()


@router.post("/", response_model=TransactionRead)
async def api_create_transaction(
    payload: TransactionCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        tx = await create_transaction(db, payload, user_id=current_user["user_id"])
        return tx
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except PermissionError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(
            status_code=500,
            detail="Failed to create transaction",
        ) from exc


@router.post("/bulk", response_model=list[TransactionRead])
async def api_create_transactions_bulk(
    payload: TransactionBulkCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    try:
        rows = await create_transactions_bulk(db, payload, user_id=current_user["user_id"])
        return [TransactionRead.model_validate(tx) for tx in rows]
    except PermissionError as exc:
        raise HTTPException(status_code=403, detail=str(exc)) from exc
    except Exception as exc:
        raise HTTPException(
            status_code=500,
            detail="Failed to import transactions",
        ) from exc


@router.get("/", response_model=TransactionListResponse)
async def api_list_transactions(
    limit: int = Query(default=100, ge=1, le=500),
    page: int = Query(default=1, ge=1),
    month: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}$"),
    start_date: date | None = Query(default=None),
    end_date: date | None = Query(default=None),
    category: str | None = Query(default=None),
    account_id: str | None = Query(default=None),
    tx_type: Literal["income", "expense", "transfer"] | None = Query(default=None, alias="type"),
    search: str | None = Query(default=None),
    sort: Literal["date", "amount", "category"] = Query(default="date"),
    sort_dir: Literal["asc", "desc"] = Query(default="desc"),
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    transactions, total_count = await list_transactions(
        db,
        user_id=current_user["user_id"],
        limit=limit,
        page=page,
        month=month,
        start_date=start_date,
        end_date=end_date,
        category=category,
        account_id=account_id,
        tx_type=tx_type,
        search=search,
        sort=sort,
        sort_dir=sort_dir,
    )
    return {
        "transactions": [TransactionRead.model_validate(tx) for tx in transactions],
        "total_count": total_count,
        "page": page,
        "limit": limit,
    }


@router.delete("/current", status_code=204)
async def api_reset_transactions(
    month: str | None = Query(default=None, pattern=r"^\d{4}-\d{2}$"),
    start_date: date | None = Query(default=None),
    end_date: date | None = Query(default=None),
    category: str | None = Query(default=None),
    account_id: str | None = Query(default=None),
    tx_type: Literal["income", "expense", "transfer"] | None = Query(default=None, alias="type"),
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    await reset_transactions(
        db,
        user_id=current_user["user_id"],
        month=month,
        start_date=start_date,
        end_date=end_date,
        category=category,
        account_id=account_id,
        tx_type=tx_type,
    )


@router.patch("/{transaction_id}", response_model=TransactionRead)
async def api_update_transaction(
    transaction_id: str,
    payload: TransactionUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    tx = await update_transaction(
        db,
        user_id=current_user["user_id"],
        transaction_id=transaction_id,
        payload=payload,
    )
    if tx is None:
        raise HTTPException(status_code=404, detail="Transaction not found")
    return tx


@router.delete("/{transaction_id}", status_code=204)
async def api_delete_transaction(
    transaction_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
):
    deleted = await delete_transaction(
        db,
        user_id=current_user["user_id"],
        transaction_id=transaction_id,
    )
    if not deleted:
        raise HTTPException(status_code=404, detail="Transaction not found")

File: src/api/transactions.ts
import client from './client'
import type { Transaction } from '@/components/index'

type TransactionApiModel = {
  id: string
  account_id?: string
  loan_id?: string | null
  amount: number
  merchant_name?: string | null
  description?: string | null
  category?: string | null
  date: string
  transaction_date?: string
  is_manual?: boolean
  created_at?: string
}

type TransactionListApiModel = {
  transactions: TransactionApiModel[]
  total_count: number
  page: number
  limit: number
}

export type TransactionQueryParams = {
  limit?: number
  page?: number
  month?: string
  start_date?: string
  end_date?: string
  category?: string
  account_id?: string
  type?: 'income' | 'expense' | 'transfer'
  search?: string
  sort?: 'date' | 'amount' | 'category'
  sort_dir?: 'asc' | 'desc'
}

export type TransactionListResult = {
  transactions: Transaction[]
  totalCount: number
  page: number
  limit: number
}

function mapTransaction(model: TransactionApiModel): Transaction {
  return {
    id: model.id,
    account_id: model.account_id ?? '',
    loan_id: model.loan_id ?? null,
    amount: model.amount,
    merchant_name: model.merchant_name ?? null,
    description: model.description ?? null,
    category: model.category ?? null,
    transaction_date: model.transaction_date ?? model.date,
    is_manual: Boolean(model.is_manual),
    created_at: model.created_at ?? new Date().toISOString(),
  }
}

export const fetchTransactions = async (params: TransactionQueryParams = {}): Promise<TransactionListResult> => {
  const { data } = await client.get<TransactionListApiModel>('/transactions/', {
    params: {
      limit: params.limit ?? 100,
      page: params.page ?? 1,
      month: params.month,
      start_date: params.start_date,
      end_date: params.end_date,
      category: params.category,
      account_id: params.account_id,
      type: params.type,
      search: params.search,
      sort: params.sort,
      sort_dir: params.sort_dir,
    },
  })

  return {
    transactions: data.transactions.map(mapTransaction),
    totalCount: data.total_count,
    page: data.page,
    limit: data.limit,
  }
}

export const createTransaction = async (payload: {
  account_id?: string | null
  loan_id?: string | null
  amount: number
  description?: string
  merchant_name?: string
  category?: string
  date: string
  notes?: string
  is_manual?: boolean
}): Promise<Transaction> => {
  const { data } = await client.post<Transaction>('/transactions/', payload)
  return data
}

export const updateTransaction = async (
  transactionId: string,
  payload: {
    amount?: number
    description?: string
    merchant_name?: string
    category?: string
    date?: string
    notes?: string
    loan_id?: string | null
  },
): Promise<Transaction> => {
  const { data } = await client.patch<Transaction>(`/transactions/${transactionId}`, payload)
  return data
}

export const deleteTransaction = async (transactionId: string): Promise<void> => {
  await client.delete(`/transactions/${transactionId}`)
}

export const resetTransactions = async (params: {
  month?: string
  start_date?: string
  end_date?: string
  category?: string
  account_id?: string
  type?: 'income' | 'expense' | 'transfer'
}): Promise<void> => {
  await client.delete('/transactions/current', { params })
}

export const bulkImportTransactions = async (payload: {
  account_id: string
  items: Array<{
    amount: number
    description?: string
    merchant_name?: string
    category?: string
    date?: string
    notes?: string
    tx_type?: 'income' | 'expense'
  }>
}): Promise<Transaction[]> => {
  const { data } = await client.post<Transaction[]>('/transactions/bulk', payload)
  return data
}
